File: 45_Hello/ruby/hello.rb
class Hello
  def start
    puts  "HELLO.  MY NAME IS CREATIVE COMPUTER.\n\n"
    print "WHAT'S YOUR NAME? "
    user_name = gets.chomp!

    ask_enjoy_question(user_name)

    ask_question_loop(user_name)

    isHonest = ask_for_fee(user_name)
    
    if isHonest
      happy_goodbye(user_name)
    else
      unhappy_goodbye(user_name)
    end
  end
  private
    def get_yes_or_no
      msg = gets.chomp!
      if msg.upcase() == "YES"
          return true, true, msg
      elsif msg.upcase() == "NO"
          return true, false, msg
      else
          return false, false, msg
      end
    end

    def ask_enjoy_question user_name
      print "\nHI THERE, #{user_name}, ARE YOU ENJOYING YOURSELF HERE? "

      while true
        valid, value, msg = get_yes_or_no()

        if valid
          if value
            puts "\nI'M GLAD TO HEAR THAT, #{user_name}."
            break
          else
            puts "\nOH, I'M SORRY TO HEAR THAT, #{user_name}. MAYBE WE CAN"
            puts "BRIGHTEN UP YOUR VISIT A BIT."
            break
          end
        else
          puts "\n#{user_name}, I DON'T UNDERSTAND YOUR ANSWER OF '#{msg}'."
          print "PLEASE ANSWER 'YES' OR 'NO'.  DO YOU LIKE IT HERE? "
        end
      end
    end

    def prompt_for_problems user_name
      puts "\nSAY, #{user_name}, I CAN SOLVE ALL KINDS OF PROBLEMS EXCEPT" 
      puts "THOSE DEALING WITH GREECE.  WHAT KIND OF PROBLEMS DO"
      print "YOU HAVE? (ANSWER SEX, HEALTH, MONEY, OR JOB) "
  
      problem_type = gets.chomp!
      return problem_type
    end

    def prompt_too_much_or_too_little
      answer = gets.chomp!
      if answer.upcase() == "TOO MUCH"
        return true, true
      elsif answer.upcase() == "TOO LITTLE"
        return true, false
      else
        return false, false
      end
    end

    def solve_sex_problem user_name
      print "\nIS YOUR PROBLEM TOO MUCH OR TOO LITTLE? "
      while true
        valid, too_much = prompt_too_much_or_too_little()
        if valid
          if too_much
            puts "\nYOU CALL THAT A PROBLEM?!!  I SHOULD HAVE SUCH PROBLEMS!"
            puts "IF IT BOTHERS YOU, #{user_name}, TAKE A COLD SHOWER."
            break
          else
            puts "\nWHY ARE YOU HERE IN SUFFERN, #{user_name}?  YOU SHOULD BE"
            puts "IN TOKYO OR NEW YORK OR AMSTERDAM OR SOMEPLACE WITH SOME"
            puts "REAL ACTION."
            break
          end
        else
          puts "\nDON'T GET ALL SHOOK, #{user_name}, JUST ANSWER THE QUESTION"
          print "WITH 'TOO MUCH' OR 'TOO LITTLE'.  WHICH IS IT? "
        end
      end
    end

    def solve_health_problem user_name
      puts "\nMY ADVICE TO YOU #{user_name} IS:"
      puts "     1.  TAKE TWO ASPRIN"
      puts "     2.  DRINK PLENTY OF FLUIDS (ORANGE JUICE, NOT BEER!)"
      puts "     3.  GO TO BED (ALONE)"
    end

    def solve_money_problem user_name
      puts "\nSORRY, #{user_name}, I'M BROKE TOO.  WHY DON'T YOU SELL"
      puts "ENCYCLOPEADIAS OR MARRY SOMEONE RICH OR STOP EATING"
      puts "SO YOU WON'T NEED SO MUCH MONEY?"
    end

    def solve_job_problem user_name
      puts "\nI CAN SYMPATHIZE WITH YOU #{user_name}.  I HAVE TO WORK"
      puts "VERY LONG HOURS FOR NO PAY -- AND SOME OF MY BOSSES"
      puts "REALLY BEAT ON MY KEYBOARD.  MY ADVICE TO YOU, #{user_name},"
      puts "IS TO OPEN A RETAIL COMPUTER STORE.  IT'S GREAT FUN."
    end

    def alert_unknown_problem_type user_name, problem_type
      puts "\nOH, #{user_name}, YOUR ANSWER OF #{problem_type} IS GREEK TO ME."
    end

    def ask_question_loop user_name
      while true
        problem_type = prompt_for_problems(user_name)
        if problem_type == "SEX"
          solve_sex_problem(user_name)
        elsif problem_type == "HEALTH"
          solve_health_problem(user_name)
        elsif problem_type == "MONEY"
          solve_money_problem(user_name)
        elsif problem_type == "JOB"
          solve_job_problem(user_name)
        else
          alert_unknown_problem_type(user_name, problem_type)
        end

        while true
          print "\nANY MORE PROBLEMS YOU WANT SOLVED, #{user_name}? "

          valid, value, msg = get_yes_or_no()
          if valid
            if value
              puts "\nWHAT KIND (SEX, MONEY, HEALTH, JOB)"
              break
            else
              return
            end
          else
            puts "\nJUST A SIMPLE 'YES' OR 'NO' PLEASE, #{user_name}."
          end
        end
      end
    end

    def ask_for_fee user_name
      puts "\nTHAT WILL BE $5.00 FOR THE ADVICE, #{user_name}."
      puts "PLEASE LEAVE THE MONEY ON THE TERMINAL."
      sleep(3)
      print "\n\nDID YOU LEAVE THE MONEY? "
  
      while true
        valid, value, msg = get_yes_or_no()
        if valid
            if value
              puts "\nHEY, #{user_name}, YOU LEFT NO MONEY AT ALL!"
              puts "YOU ARE CHEATING ME OUT OF MY HARD-EARNED LIVING."
              puts "\nWHAT A RIP OFF, #{user_name}!!!"
              return false
            else
              puts "\nTHAT'S HONEST, #{user_name}, BUT HOW DO YOU EXPECT"
              puts "ME TO GO ON WITH MY PSYCHOLOGY STUDIES IF MY PATIENTS"
              puts "DON'T PAY THEIR BILLS?"
              return true
            end
        else
          puts "\nYOUR ANSWER OF '#{msg}' CONFUSES ME, #{user_name}."
          print "PLEASE RESPOND WITH 'YES' or 'NO'. "
        end
      end
    end

    def unhappy_goodbye user_name
      puts "\nTAKE A WALK, #{user_name}.\n\n"
    end
  
    def happy_goodbye user_name
      puts "\nNICE MEETING YOU, #{user_name}, HAVE A NICE DAY."
    end
end

if __FILE__ == $0
  hello = Hello.new
  hello.start()
end

File: 45_Hello/lua/hello.lua
-- HELLO
--
-- Converted from BASIC to Lua by Recanman

local function tab(space)
    local str = ""

    for _ = space, 1, -1 do
        str = str .. " "
    end

    return str
end

-- reused from Bagels.lua
function getInput(prompt)
    io.write(prompt)
    io.flush()
    local input = io.read("l")
    if not input then  --- test for EOF
        print("GOODBYE")
        os.exit(0)
    end
    return input
end

print(tab(33) .. "HELLO\n")
print(tab(15) .. "CREATIVE COMPUTING  MORRISTOWN, NEW JERSEY\n")
print("\n")
print("\n")
print("\n")

print("HELLO.  MY NAME IS CREATIVE COMPUTER.\n")
print("\n")
print("\n")

print("WHAT'S YOUR NAME")
local ns = getInput("? ")

print("\n")
print("HI THERE, " .. ns .. ", ARE YOU ENJOYING YOURSELF HERE")

while true do
    local bs = getInput("? ")
    print("\n")
    if bs == "YES" then
        print("I'M GLAD TO HEAR THAT, " .. ns .. ".\n")
        print("\n")
        break
    elseif bs == "NO" then
        print("OH, I'M SORRY TO HEAR THAT, " .. ns .. ". MAYBE WE CAN\n")
        print("BRIGHTEN UP YOUR VISIT A BIT.\n")
        break
    else
        print("PLEASE ANSWER 'YES' OR 'NO'.  DO YOU LIKE IT HERE")
    end
end

local function main()
    print("\n")
    print("SAY, " .. ns .. ", I CAN SOLVED ALL KINDS OF PROBLEMS EXCEPT\n")
    print("THOSE DEALING WITH GREECE.  WHAT KIND OF PROBLEMS DO\n")
    print("YOU HAVE (ANSWER SEX, HEALTH, MONEY, OR JOB)")

    while true do
        local cs = getInput("? ")
        print("\n")

        if cs ~= "SEX" and cs ~= "HEALTH" and cs ~= "MONEY" and cs ~= "JOB" then
            print("OH, " .. ns .. ", YOUR ANSWER OF " .. cs .. " IS GREEK TO ME.\n")
        elseif cs == "JOB" then
            print("I CAN SYMPATHIZE WITH YOU " .. ns .. ".  I HAVE TO WORK\n")
            print("VERY LONG HOURS FOR NO PAY -- AND SOME OF MY BOSSES\n")
            print("REALLY BEAT ON MY KEYBOARD.  MY ADVICE TO YOU, " .. ns .. ",\n")
            print("IS TO OPEN A RETAIL COMPUTER STORE.  IT'S GREAT FUN.\n")
        elseif cs == "MONEY" then
            print("SORRY, " .. ns .. ", I'M BROKE TOO.  WHY DON'T YOU SELL\n")
            print("ENCYCLOPEADIAS OR MARRY SOMEONE RICH OR STOP EATING\n")
            print("SO YOU WON'T NEED SO MUCH MONEY?\n")
        elseif cs == "HEALTH" then
            print("MY ADVICE TO YOU " .. ns .. " IS:\n")
            print(tab(5) .. "1.  TAKE TWO ASPRIN\n")
            print(tab(5) .. "2.  DRINK PLENTY OF FLUIDS (ORANGE JUICE, NOT BEER!)\n")
            print(tab(5) .. "3.  GO TO BED (ALONE)\n")
        elseif cs == "SEX" then
            print("IS YOUR PROBLEM TOO MUCH OR TOO LITTLE")

            while true do
                local ds = getInput("? ")
                print("\n")

                if ds == "TOO MUCH" then
                    print("YOU CALL THAT A PROBLEM?!!  I SHOULD HAVE SUCH PROBLEMS!\n")
                    print("IF IT BOTHERS YOU, " .. ns .. ", TAKE A COLD SHOWER.\n")
                    break
                elseif ds == "TOO LITTLE" then
                    print("WHY ARE YOU HERE IN SUFFERN, " .. ns .. "?  YOU SHOULD BE\n")
                    print("IN TOKYO OR NEW YORK OR AMSTERDAM OR SOMEPLACE WITH SOME\n")
                    print("REAL ACTION.\n")
                    break
                else
                    print("DON'T GET ALL SHOOK, " .. ns .. ", JUST ANSWER THE QUESTION\n")
                    print("WITH 'TOO MUCH' OR 'TOO LITTLE'.  WHICH IS IT")
                end
            end
        end

        print("\n")
        print("ANY MORE PROBLEMS YOU WANT SOLVED, " .. ns)

        local es = getInput("? ")

        if es == "YES" then
            print("WHAT KIND (SEX, MONEY, HEALTH, JOB)")
        elseif es == "NO" then
            print("THAT WILL BE $5.00 FOR THE ADVICE, " .. ns .. ".\n")
            print("PLEASE LEAVE THE MONEY ON THE TERMINAL.\n")
            print("\n")
            print("\n")
            print("\n")

            while true do
                print("DID YOU LEAVE THE MONEY")

                local gs = getInput("? ")
                print("\n")

                if gs == "YES" then
                    print("HEY, " .. ns .. "??? YOU LEFT NO MONEY AT ALL!\n")
                    print("YOU ARE CHEATING ME OUT OF MY HARD-EARNED LIVING.\n")
                    print("\n")
                    print("WHAT A RIP OFF, " .. ns .. "!!!\n")
                    print("\n")
                    break
                elseif gs == "NO" then
                    print("THAT'S HONEST, " .. ns .. ", BUT HOW DO YOU EXPECT\n")
                    print("ME TO GO ON WITH MY PSYCHOLOGY STUDIES IF MY PATIENT\n")
                    print("DON'T PAY THEIR BILLS?\n")
                    break
                else
                    print("YOUR ANSWER OF '" .. gs .. "' CONFUSES ME, " .. ns .. ".\n")
                    print("PLEASE RESPOND WITH 'YES' OR 'NO'.\n")
                end
            end

            break
        end
    end

    print("\n")
    print("TAKE A WALK, " .. ns .. ".\n")
    print("\n")
    print("\n")
end

main()